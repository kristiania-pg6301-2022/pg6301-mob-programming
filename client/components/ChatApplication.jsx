import React, { useState } from "react";

export function ChatApplication({ messages, onNewMessage }) {
  const [message, setMessage] = useState("");
  function handleSubmit(e) {
    e.preventDefault();
    onNewMessage(message);
    setMessage("");
  }
  return (
    <>
      <header>Kristiania Chat</header>
      <main>
        {messages.map(({ message, userInfo }, index) => (
          <div key={index}>
            <strong>{userInfo}:</strong> {message}
          </div>
        ))}
      </main>
      <footer>
        <form onSubmit={handleSubmit}>
          <input
            autoFocus={true}
            value={message}
            onChange={(e) => setMessage(e.target.value)}
          />
          <button>Send</button>
        </form>
      </footer>
    </>
  );
}
