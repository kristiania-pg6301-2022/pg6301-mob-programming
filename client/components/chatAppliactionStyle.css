body {
    margin: 0;
}

#app {
    display: grid;
    grid-template-rows: auto 1fr auto;
    height: 100vh;
}

main {
    background-color: aliceblue;
}

footer form {
    display: flex;
    align-items: stretch;
    flex-direction: row;
}

footer form input {
    flex: 1;
}
