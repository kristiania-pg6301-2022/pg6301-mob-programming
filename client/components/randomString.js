export default function randomString(length) {
  const symbols =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  let result = "";
  for (let i = 0; i < length; i++) {
    result = symbols.charAt(Math.floor(Math.random() * symbols.length));
  }
  return result;
}
