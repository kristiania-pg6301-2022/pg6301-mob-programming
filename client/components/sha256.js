export async function sha256(string) {
  const binaryHash = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder("utf-8").encode(string)
  );
  return btoa(String.fromCharCode.apply(null, new Uint8Array(binaryHash)))
    .split("=")[0]
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}
