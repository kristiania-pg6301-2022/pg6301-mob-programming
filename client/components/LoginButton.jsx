import randomString from "./randomString";
import { sha256 } from "./sha256";
import React from "react";

export function LoginButton({ config, label, provider }) {
  async function handleLogin() {
    const {
      authorization_endpoint,
      response_type,
      scope,
      client_id,
      code_challenge_method,
    } = config[provider];

    const state = randomString(50);
    window.sessionStorage.setItem("expected_state", state);

    const parameters = {
      response_type,
      response_mode: "fragment",
      client_id,
      state,
      scope,
      redirect_uri: `${window.location.origin}/login/${provider}/callback`,
    };

    if (code_challenge_method) {
      const state = randomString(50);
      window.sessionStorage.setItem("authorization_state", state);
      const code_verifier = randomString(50);
      window.sessionStorage.setItem("code_verifier", code_verifier);
      parameters.code_challenge_method = code_challenge_method;
      parameters.code_challenge = await sha256(code_verifier);
      parameters.domain_hint = "egms.no";
    }

    window.location.href = `${authorization_endpoint}?${new URLSearchParams(
      parameters
    )}`;
  }

  return (
    <div>
      <button onClick={handleLogin}>{label}</button>
    </div>
  );
}
