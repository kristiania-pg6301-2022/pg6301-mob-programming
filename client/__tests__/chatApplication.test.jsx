import { ChatApplication } from "../components/ChatApplication";
import ReactDOM from "react-dom";
import React from "react";
import { Simulate } from "react-dom/test-utils";

describe("Chat application", () => {
  it("should show chats", () => {
    const chats = [
      { user: "test user", message: "message 1" },
      { user: "other user", message: "message 2" },
    ];

    const element = document.createElement("div");
    ReactDOM.render(<ChatApplication messages={chats} />, element);
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should submit new chat", () => {
    const element = document.createElement("div");
    const onNewMessage = jest.fn();
    ReactDOM.render(
      <ChatApplication messages={[]} onNewMessage={onNewMessage} />,
      element
    );
    Simulate.change(element.querySelector("footer input"), {
      target: { value: "new message" },
    });
    Simulate.submit(element.querySelector("footer form"));
    expect(onNewMessage).toBeCalledWith("new message");
  });
});
