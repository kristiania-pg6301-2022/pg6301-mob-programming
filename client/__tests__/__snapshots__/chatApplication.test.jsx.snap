// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat application should show chats 1`] = `"<header>Kristiania Chat</header><main><div><strong>:</strong> message 1</div><div><strong>:</strong> message 2</div></main><footer><form><input value=\\"\\"><button>Send</button></form></footer>"`;
